### ISSUE TITLE
Pressing "save" more than once discards all changes

### ISSUE THREAD

**Issue by rasitayaz** [tralph3/Steam-Metadata-Editor#37]
In SteamOS, if you edit and save more than one game (or edit the same game and save again), none of the edits will be applied on Steam. After pressing the "save" button, the app needs to be restarted to make further changes because pressing it again just discards every edit made in that session.

**Comment by AlJoelson** [comment]
This issue also occurs on Windows.

**Comment by sonic2kk** [comment]
This appears to be a a result of invalid data being written to the VDF file, which occurs when trying to save twice. If you press the save button twice, then close and attempt to re-open SME, it will give an error like this:

```
Traceback (most recent call last):
  File "/home/user/Programming/Steam-Metadata-Editor/src/main.py", line 45, in <module>
    main()
  File "/home/user/Programming/Steam-Metadata-Editor/src/main.py", line 34, in main
    mainWindow = MainWindow()
                 ^^^^^^^^^^^^
  File "/home/user/Programming/Steam-Metadata-Editor/src/gui/main_window.py", line 91, in __init__
    self.create_main_window()
  File "/home/user/Programming/Steam-Metadata-Editor/src/gui/main_window.py", line 110, in create_main_window
    self.appinfo = Appinfo(self.vdf_path)
                   ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/Programming/Steam-Metadata-Editor/src/appinfo.py", line 58, in __init__
    self.parsedAppInfo = self.read_all_apps()
                         ^^^^^^^^^^^^^^^^^^^^
  File "/home/user/Programming/Steam-Metadata-Editor/src/appinfo.py", line 172, in read_all_apps
    app["sections"] = self.parse_subsections()
                      ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/Programming/Steam-Metadata-Editor/src/appinfo.py", line 105, in parse_subsections
    value = value_parsers[value_type]()
            ~~~~~~~~~~~~~^^^^^^^^^^^^
KeyError: 172
```

When you open Steam, it simply resolves this corruption, making it seem like the edits are cleared.

**Comment by tralph3** [comment]
I have figured out what the issue is. When you press save, the modified data gets encoded into its binary form, and the original data is replaced in a bytearray that contains the entire appinfo.vdf.

In order to know *where* in that entire bytearray the data for this particular app is located, I grab the binary data for the app's header, and perform a search. The search either returns an index, and I drop the new data, or it doesn't find anything, and I drop the new data at the end of the file.

Now here's the problem. The app data has now been updated, meaning the header has changed too. If you press save again, the search will be performed again, but with the previous state of the header. It won't find anything now. And so the data gets appended at the end, and Steam must see that the info is repeated twice, and it's probably just invalidating the cache and recreating it.

So I need to investigate how to fix that. I can probably make it so it doesn't search for the entire header, and just use the appId, although that introduces the possibility that somewhere in the file that is exact number is repeated, and it will make a mess.

I could probably search for the ID prepended by '0x08' which is the code for the end of a dictionary. The only issue with that is that the very first app wouldn't have a dictionary ending before. But I think the very first one is some internal Steam stuff, and not a real game, so it's not really an issue, although it does feel a bit dirty to me.

Another possibility would be to update the header stored in the variables, but I would need to look into it since I don't quite recall how it's structured.
